Invalid C codegen `refc` with generic types containing gc memory

### Nim Version

```
Nim Compiler Version 2.0.14 [Linux: amd64]
Compiled at 2025-04-03
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: bf4de6a394e040d9810cba8c69fb2829ff04dcc6
active boot switches: -d:release
```

```
Nim Compiler Version 2.2.3 [Linux: amd64]
Compiled at 2025-04-03
Copyright (c) 2006-2025 by Andreas Rumpf

git hash: a2a6565e23229b7415d28ef2859c6c593db49989
active boot switches: -d:release
```

```
Nim Compiler Version 2.3.1 [Linux: amd64]
Compiled at 2025-04-03
Copyright (c) 2006-2025 by Andreas Rumpf

git hash: 4352fa2ef0cbba953d9a90b90873e8dd0364b72e
active boot switches: -d:release
```

```
gcc (Debian 14.2.0-19) 14.2.0
Copyright (C) 2024 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

### Description

The main module:
```nim
import "."/[r, v]
type M = ref object
type E = distinct int
method j(_: M) {.base.} = u()
discard p[E]()
```

`r.nim`:
```nim
import "."/v
proc u*() =
  iterator c(): int {.closure.} = raiseAssert ""
  discard A[int]()
  let _ = c
```

`v.nim`:
```nim
type
  B = ref object of W
  W = object of RootObj
    y: proc(_: int)
    n: B
    k: iterator(): int {.closure.}
  S*[T] = ref object of B
    r: T
  A*[T] = ref object of S[T]

proc p*[T](): S[T] = S[T]()
iterator t(): int {.closure.} = discard
```

### Current Output

```text
Hint: used config file '/tmp/nimdevel/config/nim.cfg' [Conf]
Hint: used config file '/tmp/nimdevel/config/config.nims' [Conf]
..........................................................................
/tmp/v.nim(12, 10) Hint: 't' is declared but not used [XDeclaredButNotUsed]
CC: nimdevel/lib/system/exceptions.nim
CC: nimdevel/lib/std/private/digitsutils.nim
CC: nimdevel/lib/std/assertions.nim
CC: nimdevel/lib/system/dollars.nim
CC: nimdevel/lib/std/typedthreads.nim
CC: nimdevel/lib/pure/collections/sharedlist.nim
CC: nimdevel/lib/system.nim
CC: v.nim
CC: r.nim
CC: f.nim
/tmp/tmp.BBAWLItgpC/@mf.nim.c: In function ‘Marker_tyRef__NBQRyu0xvXLho06pe9bfbog’:
/tmp/tmp.BBAWLItgpC/@mf.nim.c:115:25: error: invalid use of incomplete typedef ‘tyObject_ScolonObjectType___PsOs7wPZadaq8vGxWAk0CA’
  115 | nimGCvisit(((void*) (*a).Sup.Sup.y.ClE_0), op);
      |                         ^
/tmp/tmp.BBAWLItgpC/@mf.nim.c:116:25: error: invalid use of incomplete typedef ‘tyObject_ScolonObjectType___PsOs7wPZadaq8vGxWAk0CA’
  116 | nimGCvisit(((void*) (*a).Sup.Sup.n), op);
      |                         ^
/tmp/tmp.BBAWLItgpC/@mf.nim.c:117:25: error: invalid use of incomplete typedef ‘tyObject_ScolonObjectType___PsOs7wPZadaq8vGxWAk0CA’
  117 | nimGCvisit(((void*) (*a).Sup.Sup.k.ClE_0), op);
      |                         ^
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3 -pthread -fno-omit-frame-pointer   -I/tmp/nimdevel/lib -I/tmp -o /tmp/tmp.BBAWLItgpC/@mf.nim.c.o /tmp/tmp.BBAWLItgpC/@mf.nim.c' failed with exit code: 1
```

### Expected Output

```text

```

### Known Workarounds

_No response_

### Additional Information

_No response_

--- a/compiler/ccgtypes.nim
+++ b/compiler/ccgtypes.nim
@@ -2003,7 +2003,7 @@ proc genTypeInfoV1(m: BModule; t: PType; info: TLineInfo): Rope =
     if m.config.selectedGC in {gcMarkAndSweep, gcRefc, gcGo}:
       # it may not be used in other places except in `genTraverseProc`,
       # we have to generate a typedesc for this case, not a weak one
-      discard getTypeDesc(m, origType.base)
+      discard getTypeDesc(m, origType.last)
       let markerProc = genTraverseProc(m, origType, sig)
       m.s[cfsTypeInit3].addFieldAssignment(tiNameForHcr(m, result), "marker", markerProc)
   of tyPtr, tyRange, tyUncheckedArray: genTypeInfoAux(m, t, t, result, info)

--- a/compiler/ccgtypes.nim
+++ b/compiler/ccgtypes.nim
@@ -2001,8 +2001,8 @@ proc genTypeInfoV1(m: BModule; t: PType; info: TLineInfo): Rope =
   of tyRef:
     genTypeInfoAux(m, t, t, result, info)
     if m.config.selectedGC in {gcMarkAndSweep, gcRefc, gcGo}:
-      # it may not be used in other places exccept in `genTraverseProc`,
-      # we have to generate a typedesc for this case
+      # it may not be used in other places except in `genTraverseProc`,
+      # we have to generate a typedesc for this case, not a weak one
       discard getTypeDesc(m, origType.base)
       let markerProc = genTraverseProc(m, origType, sig)
       m.s[cfsTypeInit3].addFieldAssignment(tiNameForHcr(m, result), "marker", markerProc)

--- a/compiler/ccgtypes.nim
+++ b/compiler/ccgtypes.nim
@@ -2001,6 +2001,9 @@ proc genTypeInfoV1(m: BModule; t: PType; info: TLineInfo): Rope =
   of tyRef:
     genTypeInfoAux(m, t, t, result, info)
     if m.config.selectedGC in {gcMarkAndSweep, gcRefc, gcGo}:
+      # it may not be used in other places exccept in `genTraverseProc`,
+      # we have to generate a typedesc for this case
+      discard getTypeDesc(m, origType.base)
       let markerProc = genTraverseProc(m, origType, sig)
       m.s[cfsTypeInit3].addFieldAssignment(tiNameForHcr(m, result), "marker", markerProc)
   of tyPtr, tyRange, tyUncheckedArray: genTypeInfoAux(m, t, t, result, info)

--- a/tests/refc/m24844.nim
+++ b/tests/refc/m24844.nim
@@ -1,0 +1,8 @@
+type
+  S*[T] = ref object of RootObj
+    k: string
+  A*[T] = ref object of S[T]
+
+proc p*[T](): S[T] = S[T]()
+proc u*() = discard A[int]()
+discard A[int]()

--- a/tests/refc/t24844.nim
+++ b/tests/refc/t24844.nim
@@ -1,0 +1,11 @@
+discard """
+  matrix: "--mm:refc; --mm:arc"
+  joinable: false
+"""
+
+import m24844
+
+u()
+
+type E = distinct int
+discard p[E]()

